pkg/sql/schemachanger/eav/tree.go

```diff
@@ -0,0 +1,271 @@
+	}
+	// Index everything by all of the attributes. This serves as the primary
+	// index.
+	t.dims[0] = indexSpec{s: sc}
+	dims := t.dims[1:]
+	for i, attrs := range indexes {
+		m := MakeOrdinalSetWithAttributes(attrs)
+		dims[i] = indexSpec{mask: m, attrs: attrs, s: sc}
+	}
+	return t
+}
+
+// Insert inserts an entity.
+func (t *Tree) Insert(e Entity) (removed Entity) {
+	removedItem := t.t.ReplaceOrInsert(&containerItem{
+		Entity:    e,
+		indexSpec: &t.dims[0]},
+	)
+	if removedItem != nil && !Equal(t.schema, removedItem.(Entity), e) {
+		panic("here")
+	}
+	dims := t.dims[1:]
+	for i := range dims {
+		if g := t.t.ReplaceOrInsert(&containerItem{
+			Entity:    e,
+			indexSpec: &dims[i],
+		}); (removedItem == nil) != (g == nil) {
+			panic(errors.AssertionFailedf(
+				"expected to remove the item each time: %v %v", removedItem, g,
+			))
+		}
+	}
+	if removedItem != nil {
+		return removedItem.(*containerItem).Entity
+	}
+	return nil
+}
+
+type indexSpec struct {
+	s     Schema
+	mask  OrdinalSet
+	attrs []Attribute
+}
+
+// Iterate will iterate the containers which match the specified values.
+func (t *Tree) Iterate(where Values, f EntityIterator) (err error) {
+	var all, nils, nonNils OrdinalSet
+	{
+		all = where.Attributes()
+		all.ForEach(t.schema, func(a Attribute) (wantMore bool) {
+			if where.Get(a) == nil {
+				nils = nils.Add(a.Ordinal())
+			}
+			return true
+		})
+		nonNils = all.Without(nils)
+	}
+
+	idx, toCheck := t.chooseIndex(all)
+	from, to := getValuesItems(idx, where, all)
+	defer putValuesItems(from, to)
+	t.t.AscendRange(from, to, func(i btree.Item) (wantMore bool) {
+		c := i.(*containerItem)
+		// We want to skip items which do not have values set for
+		// all members of the where clause or which have values set
+		// for attributes where we explicitly do not want them.
+		if cAttrs := c.Attributes(); nonNils.Without(cAttrs) != 0 ||
+			nils.Intersection(cAttrs) != 0 {
+			return true
+		}
+		var failed bool
+		toCheck.ForEach(t.schema, func(a Attribute) (wantMore bool) {
+			_, eq := CompareOn(a, c.Entity, where)
+			failed = !eq
+			return !failed
+		})
+		if !failed {
+			err = f(c.Entity)
+		}
+		return err == nil
+	})
+	if iterutil.Done(err) {
+		err = nil
+	}
+	return err
+}
+
+// item is implemented by all members of the tree as well as by the valuesItem
+// used at query time.
+type item interface {
+	btree.Item
+	getIndexSpec() *indexSpec
+	Entity
+}
+
+var _ item = (*containerItem)(nil)
+var _ item = (*valuesItem)(nil)
+
+// chooseIndex chooses an index which has a prefix with the highest number of
+// attributes which overlap with m. It also returns the ordinals of the
+// attributes which are not covered by the index prefix.
+//
+// TODO(ajwerner): Consider something about selectivity by tracking
+// the number of entries under each index (i.e. which have non-NULL values)
+// for the given dimension.
+func (t *Tree) chooseIndex(m OrdinalSet) (_ *indexSpec, toCheck OrdinalSet) {
```

Comment:
mini database 👶

Reply:
In seriousness, I don't understand the motivation for this and it seems complicated! Complicated is fine if warranted. Why do we need to have such an advanced data structure for this library? Would a basic search structure work as well, or are we defeating some kind of terrible O() with this that is not possible to work around in a different way?